File: 06/perl/main.pl
#!/usr/bin/env perl

use strict;
use warnings;
use utf8;
use feature qw( say );

use Carp;

=head1 DESCRIPTION

This is the solution for part 1 and 2 of day 6.

=head2 main

Man method to run the program

=cut

sub main {
    my $file = $ARGV[0] || croak 'No input file';

    open( my $fh, '<', $file ) || croak 'Could not open file: ' . $!;
    my @lines = <$fh>;
    close( $fh );

    my %orbit   = ();
    my %backref = ();

    # Map all orbits and keep a backref to where the current orbit came from.
    foreach my $relation ( @lines ) {
        chomp $relation;

        my ( $lhs, $rhs ) = split( /\)/x, $relation );
        $orbit{$lhs}{$rhs} = 1;
        $backref{$rhs} = $lhs;
    }

    my %distances = ();
    my $total     = 0;

    # Iterate through all orbits and traverse them back to 'COM' to get a hash
    # back with each planet as key and the distance to 'COM' as value.
    find_distances_from( 'COM', 0, \%orbit, \%distances );

    # Sum all the distances.
    map { $total += $distances{$_} } keys %distances;

    # Part one result.
    say "Total orbits: $total";

    # Store all backreferences from 'YOU' and 'SAN' in one hash each with their
    # backrefs as key.
    my $backref_src = get_backrefs_from( 'YOU', \%backref, {} );
    my $backref_dst = get_backrefs_from( 'SAN', \%backref, {} );

    # Since backref only contains one planet we know that the number of planets
    # seen is the number of steps back to 'COM'.
    my $src_distances_from_com = scalar keys %$backref_src;
    my $dst_distances_from_com = scalar keys %$backref_dst;

    # Now we just need to find all the common planets (where the orbital courses
    # meet)
    my @common_keys = ();
    foreach my $key ( keys %$backref_src ) {
        push @common_keys, $key if exists $backref_dst->{$key};
    }

    # Check which common planet is the furthest out from 'COM' (and thus closest
    # to us).
    my $nearest_shared_distance = 0;
    foreach my $key ( @common_keys ) {
        $nearest_shared_distance = $distances{$key}
          if $distances{$key} > $nearest_shared_distance || $nearest_shared_distance == 0;
    }

    # Take the delta of the nearest sahred and our planets and add them together
    # to get the distance between 'YOU' and 'SAN'.
    my $delta_src = $src_distances_from_com - $nearest_shared_distance;
    my $delta_dst = $dst_distances_from_com - $nearest_shared_distance;
    my $delta     = $delta_src + $delta_dst - 2;                       # Remove the planets themself.

    # Part two result.
    say "Minimum orbits to move from 'YOU' to 'SAN': $delta";

    return 1;
}

=head2 get_backrefs_from

Recursive function to get a flat hash of all the planets behind $planet.

=cut

sub get_backrefs_from {
    my ( $planet, $backref, $list ) = @_;

    my $backrefed_planet = $backref->{$planet};

    return $list unless $backrefed_planet;

    $list->{$backrefed_planet} = 1;

    {
        no warnings 'recursion';
        get_backrefs_from( $backrefed_planet, $backref, $list );
    }

    return $list;
}

=head2 find_distances_from

Recursive function to get a flat hash of the distance to inputted planet for
each child.

=cut

sub find_distances_from {
    my ( $planet, $distance, $orbit, $distances ) = @_;

    $distances->{$planet} = $distance;

    my $next_distnace = ++$distance;

    foreach my $child ( keys %{ $orbit->{$planet} } ) {
        {
            no warnings 'recursion';
            find_distances_from( $child, $next_distnace, $orbit, $distances );
        };
    }

    return $distances;
}

main();

File: 06/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strings"
)

func main() {
	orbits := readFile()

	f, b := parseMap(orbits)
	d := map[string]int{}

	findDistanceFrom("COM", 0, f, d)

	totalOrbits := 0
	for _, distance := range d {
		totalOrbits += distance
	}

	// Part 1 solution
	fmt.Println("Total orbis", totalOrbits)

	var (
		youBackrefs           = backrefsFor("YOU", b, map[string]struct{}{})
		sanBackrefs           = backrefsFor("SAN", b, map[string]struct{}{})
		commonKeys            = []string{}
		nearestSharedDistance = 0
	)

	// Check common keys found as backrefs from both 'YOU' and 'SAN'.
	for k := range youBackrefs {
		if _, ok := sanBackrefs[k]; ok {
			commonKeys = append(commonKeys, k)
		}
	}

	// Iterate over the common ones and find the distance to the one closest to
	// you/farthest 'COM'.
	for _, k := range commonKeys {
		if nearestSharedDistance == 0 || d[k] > nearestSharedDistance {
			nearestSharedDistance = d[k]
		}
	}

	// Add the distance between 'YOU' and the closest shared + the distance
	// between 'SAN' and the closest shared one to get a delta.
	srcMinusNearest := len(youBackrefs) - nearestSharedDistance
	dstMinusNearest := len(sanBackrefs) - nearestSharedDistance
	delta := srcMinusNearest + dstMinusNearest - 2 // Remove 'YOU' and 'SAN'.

	// Part 2 solution
	fmt.Println("Minimum orbits to move between 'YOU' and 'SAN':", delta)
}

func backrefsFor(planet string, backrefs map[string]string, m map[string]struct{}) map[string]struct{} {
	backref, ok := backrefs[planet]
	if !ok {
		return m
	}

	m[backref] = struct{}{}

	return backrefsFor(backref, backrefs, m)
}

func findDistanceFrom(planet string, distance int, orbits map[string][]string, store map[string]int) {
	store[planet] = distance
	nextDistance := distance + 1

	f, ok := orbits[planet]
	if !ok {
		return
	}

	for _, p := range f {
		findDistanceFrom(p, nextDistance, orbits, store)
	}
}

func parseMap(orbits []string) (map[string][]string, map[string]string) {
	var (
		forward  = map[string][]string{}
		backward = map[string]string{}
	)

	for _, line := range orbits {
		parts := strings.Split(line, ")")
		lhs, rhs := parts[0], parts[1]

		forward[lhs] = append(forward[lhs], rhs)
		backward[rhs] = lhs
	}

	return forward, backward
}

func readFile() []string {
	if len(os.Args) < 2 {
		log.Fatal("missing file as input")
	}

	line, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		log.Fatalf("could not read file: %s", err.Error())
	}

	return strings.Split(strings.TrimSpace(string(line)), "\n")
}
